### ISSUE TITLE
Wabbajack not clearing Webview2 Instances, going out of memory

### ISSUE THREAD

**Issue by baconsplit** [wabbajack-tools/wabbajack#2459]
**Describe the bug**

Wabbajack doesn't seem to exit finished downloads and their Webview2 instances
**Additional Context**

![image](https://github.com/wabbajack-tools/wabbajack/assets/63647668/765c8945-dcd1-4039-8e2a-6822ac4faa69)

OS : Up to date Windows 10
Game : Skyrim SE
Wabbajack not located in a system folder

Wabbajack Version

3.4.0.0

Logs
Only the download finished logs.
```
00:35:41.603 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Interaction Icons Rework-52743-2-1-5-1699221003.7z. Hash: 5Ckopk7AqmU=; Size: 3,7MB/3,7MB
00:35:41.682 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 21524 - Remaining Limit: 10000
00:35:45.197 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Countess - Himbo-75164-1-1-1662946232.7z. Hash: K2ej4Cp3aos=; Size: 3,7MB/3,7MB
00:35:45.270 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 10995 - Remaining Limit: 10000
00:35:46.880 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Save Unbaker-85565-1-0-4-1677453959.7z. Hash: 2bCk5RcqK24=; Size: 3,7MB/3,7MB
00:35:46.971 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 11599 - Remaining Limit: 10000
00:35:51.780 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading 2K Falmer ear-69258-1-0-1654589777.7z. Hash: 3pgulUzL+C4=; Size: 3,7MB/3,7MB
00:35:51.971 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 135232 - Remaining Limit: 10000
00:35:55.072 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Crafting Recipe Distributor-52276-2-1-0-1665595840.7z. Hash: J1ms7al/8h0=; Size: 3,7MB/3,7MB
00:35:55.214 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 25116 - Remaining Limit: 10000
00:35:58.618 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Ferrum Nibenis - My Patches SE by Xtudo - HDT-SMP-89986-1-7-1684694825.7z. Hash: JV0F5C8O6wo=; Size: 3,8MB/3,8MB
00:35:58.821 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 84105 - Remaining Limit: 10000
00:36:01.108 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Blade and Blunt - A Combat Overhaul-34549-3-1-1-1690900904.7z. Hash: qDG7tZf04y0=; Size: 3,8MB/3,8MB
00:36:01.257 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 48645 - Remaining Limit: 10000
00:36:04.172 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Imperial Metal Replacer Texture-9119-1-1.7z. Hash: Y5m7wznbq9w=; Size: 3,8MB/3,8MB
00:36:04.252 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 18475 - Remaining Limit: 10000
00:36:07.027 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading SkyHUD v090B v4-463-0-90B.zip. Hash: AZdJVvOj2v0=; Size: 3,8MB/3,8MB
00:36:07.121 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 15347 - Remaining Limit: 10000
00:36:09.965 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Modpocalypse NPCs (v3) Amber Guard-61504-1-0-1641611393.7z. Hash: HgQlfHxvLOs=; Size: 3,8MB/3,8MB
00:36:10.053 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 20610 - Remaining Limit: 10000
00:36:15.117 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Dual Casting Fix-92454-1-0-1685396946.7z. Hash: T4iwS6sVFO4=; Size: 3,8MB/3,8MB
00:36:15.193 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 9176 - Remaining Limit: 10000
00:36:16.793 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Dynamic Things Alternative - Base Object Swapper-60741-0-2-2-1691407264.7z. Hash: rshXaa4lkeQ=; Size: 3,8MB/3,8MB
00:36:16.869 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 11147 - Remaining Limit: 10000
00:36:19.221 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Splashes of Storms-72115-1-3-0-1666695424.7z. Hash: Bx5WqXMnkXE=; Size: 3,8MB/3,8MB
00:36:19.325 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 23805 - Remaining Limit: 10000
00:36:23.245 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading 3BA Vanilla Armors - No Rim Lighting-59268-1-3-1661534755.zip. Hash: MbiFiSiZEzc=; Size: 3,8MB/3,8MB
00:36:23.381 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 33818 - Remaining Limit: 10000
00:36:27.107 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Animunculi - Dwarven Sphere SFX Redesigned (Aggressive Version) --87528-1-1-1679432314.7z. Hash: 2IMgNCgATlY=; Size: 3,9MB/3,9MB
00:36:27.233 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 38737 - Remaining Limit: 10000
00:36:29.458 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Show Player In Menus - French-82137-2-0-3FR03-1686098657.7z. Hash: rXAMvE+qS6Q=; Size: 3,9MB/3,9MB
00:36:29.638 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 53347 - Remaining Limit: 10000
00:36:31.863 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Smoking Torches and Candles-8607-1-76-1616249774.zip. Hash: u6ivHXs2Rck=; Size: 3,9MB/3,9MB
00:36:31.942 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 18904 - Remaining Limit: 10000
00:36:34.884 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading MergeMapper-74689-1-5-0-1683883576.7z. Hash: Z11Hnm95MRI=; Size: 3,9MB/3,9MB
00:36:34.978 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 39883 - Remaining Limit: 10000
00:36:37.768 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Simple Offence Suppression-41764-2-2-1-1665496323.7z. Hash: 6kutT698JRA=; Size: 3,9MB/3,9MB
00:36:37.840 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 14764 - Remaining Limit: 10000
00:36:40.445 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading NPC Animation Remix (OAR) - main archive-63471-1-6-2-1696597265.zip. Hash: AXQBPcAhQ3c=; Size: 3,9MB/3,9MB
00:36:40.585 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 21495 - Remaining Limit: 10000
00:36:43.395 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Narrative Gameplay Consistent Dialogue Tweaks-64667-1-9-2-1697458703.7z. Hash: TEWxlYsf2oo=; Size: 3,9MB/3,9MB
00:36:43.923 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 463658 - Remaining Limit: 10000
00:36:47.472 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Rogue Master Detector-93040-1-0-3-1686187500.7z. Hash: sMnalSzuVbY=; Size: 3,9MB/3,9MB
00:36:47.615 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 15252 - Remaining Limit: 10000
00:36:49.591 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Traveling Mage CBBE 3BA-69407-1-1-1654817844.zip. Hash: nDmP/Sv2igQ=; Size: 3,9MB/3,9MB
00:36:49.679 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 18835 - Remaining Limit: 10000
00:36:52.793 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading House of Horrors Quest Expansion - Patches-57584-1-3-0-1643397185.rar. Hash: AfiltdzK9OU=; Size: 3,9MB/3,9MB
00:36:52.912 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 30855 - Remaining Limit: 10000
00:36:56.542 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Savage Offspring SKSE-49377-2-1-0-1665587512.7z. Hash: KwZOPj5K7H0=; Size: 4MB/4MB
00:36:56.633 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 27785 - Remaining Limit: 10000
00:36:59.073 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading ALT - RowBoat - The Gislinge - 2K Texture-88403-1-0-1680509765.7z. Hash: XJyHBdJ+CH0=; Size: 4MB/4MB
00:36:59.165 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 17031 - Remaining Limit: 10000
00:37:02.308 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading 2K x 1K Version.-97385-1-0-1690910682.7z. Hash: IvcUtlRDHZk=; Size: 4MB/4MB
00:37:02.687 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 15843 - Remaining Limit: 10000
00:37:05.241 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Weightless-93796-1-2-1687714032.7z. Hash: Spw1gGanpmk=; Size: 4MB/4MB
00:37:05.331 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 32756 - Remaining Limit: 10000
00:37:09.505 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Lennys Claybrick Replacer 2K-94227-1-0-1687677394.rar. Hash: bhOaBIBuLYM=; Size: 4MB/4MB
00:37:09.583 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 18662 - Remaining Limit: 10000
00:37:12.937 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Cathedral - 3D Clovers-68793-1-1653789098.7z. Hash: Evl3kg60/u4=; Size: 4MB/4MB
00:37:13.970 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 515621 - Remaining Limit: 10000
00:37:15.924 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Better Butterflies Monarch in a TMD Jar-79332-1-0-1675655518.rar. Hash: TjM0Tp5vYEc=; Size: 4MB/4MB
00:37:16.005 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 21488 - Remaining Limit: 10000
00:37:20.018 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading DPI Scaling Fix-95740-1-0-1689242798.zip. Hash: B5CzYQ5w+r8=; Size: 4MB/4MB
00:37:20.111 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 22164 - Remaining Limit: 10000
00:37:22.686 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Skyrim Souls RE - Updated (Anniversary Edition)-27859-2-3-2-1667836222.zip. Hash: r5TXifBA6aY=; Size: 4MB/4MB
00:37:22.811 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 14773 - Remaining Limit: 10000
00:37:26.237 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading MCO Universal Support-85491-1-0-1676987105.zip. Hash: ovZQjH3UAV8=; Size: 4MB/4MB
00:37:26.318 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 20581 - Remaining Limit: 10000
00:37:28.859 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading 3D Trellis Improved - Gildergreen Plaza-75882-0-2-1664235676.7z. Hash: vIOWtZQmgmU=; Size: 4MB/4MB
00:37:28.953 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 21773 - Remaining Limit: 10000
00:37:31.743 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading ElSopa - High Quality Doll 2k-47703-1-0-1616968013.7z. Hash: mDFWMF6o27o=; Size: 4,1MB/4,1MB
00:37:31.823 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 12916 - Remaining Limit: 10000
00:37:35.146 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Mihail Talkative Dragon Priests Patch-33084-1-3Patches-1683447319.7z. Hash: rbRKmTv5Sg4=; Size: 4,1MB/4,1MB
00:37:35.223 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 19708 - Remaining Limit: 10000
00:37:37.432 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Wind Ruler Armor - HIMBO version-60893-1-2-1641086958.zip. Hash: v6YH7m49dVI=; Size: 4,1MB/4,1MB
00:37:37.535 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 14714 - Remaining Limit: 10000
00:37:40.166 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Moons And Stars - Sky Overhaul SKSE-73336-2-0-2-1665609139.7z. Hash: b0bS2uaNR+0=; Size: 4,1MB/4,1MB
00:37:40.247 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 17145 - Remaining Limit: 10000
00:37:43.184 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading Powdery Bone Meal SE-79305-v1-0-1669245531.7z. Hash: aE+xMsVVqGg=; Size: 4,1MB/4,1MB
00:37:43.261 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 17286 - Remaining Limit: 10000
00:37:46.299 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading ENB Lights For Effect Shaders-56362-2-1-0-1665397992.7z. Hash: TFnAqNG7l1o=; Size: 4,1MB/4,1MB
00:37:46.436 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 20899 - Remaining Limit: 10000
00:37:48.983 [INFO] (Wabbajack.Downloaders.DownloadDispatcher) Finished downloading ElSopa - HD Better Bloody Rags 2k-44059-1-0-1609951714.7z. Hash: WCTW1UiWwn4=; Size: 4,1MB/4,1MB
00:37:49.183 [INFO] (Wabbajack.Networking.NexusApi.ProxiedNexusApi) Nexus API call finished: 33403 - Remaining Limit: 10000
```

**Comment by baconsplit** [comment]
related issue:
https://github.com/wabbajack-tools/wabbajack/issues/2080

**Comment by Mieekaserra** [comment]
Its not closing them as a band-aid fix for #2440

**Comment by baconsplit** [comment]
Alright but this seems to also fill up my memory, but I haven't seen the crash yet as described in #2440
I have to restart wabbajack from time to time now tho.
But good, if it is a known problem and fix for something else. You can close this ticket then I guess

**Comment by Nostras** [comment]
Having the same issue. I have lots and lots of RAM and swap which causes the GPU VRAM to fill up for me instead causing Windows to behave funny.
![image](https://github.com/wabbajack-tools/wabbajack/assets/4203238/3d41cee0-4149-478a-a8e7-db3166c6de91)
![image](https://github.com/wabbajack-tools/wabbajack/assets/4203238/daef8af3-1ef2-4bdb-b8c4-8849bf34fa7d)
But yeah, it seems that this problem got reintroduced in 3.4.

**Comment by marcogruebler** [comment]
@EzioTheDeadPoet  sorry but could you please remove the "not a bug" tag, and instead maybe even suggest a date for a fix?
With this behavior actually killing a whole system (and not just simply inhibiting the use of the program itself), this is a major blocking issue and no one should use it !

**Comment by Althro** [comment]
Wabbajack is worked on by people in their free time. You are welcome to contribute a fix should you be skilled in coding, but until then this will stand.

**Comment by marcogruebler** [comment]
Understood, but until then it should not be in public usage is my point. Most ppl arent even skilled enough to troubleshoot such a misbehavior.

**Comment by Althro** [comment]
The application works for what it is intended to. The webview only occurs on non nexus premium installations. Saying something should be pulled when multiple thousand people have successful installs each day is not conducive to getting a fix in.

As stated before, you are welcome to contribute a fix but until then I would advise you do not continue down this track.

**Comment by EzioTheDeadPoet** [comment]
> @EzioTheDeadPoet sorry but could you please remove the "not a bug" tag, and instead maybe even suggest a date for a fix? With this behavior actually killing a whole system (and not just simply inhibiting the use of the program itself), this is a major blocking issue and no one should use it !

This is not a bug. 
This is a necessary INTENDED behavior to keep the program usable for especially manual installations but also the automated ones that occasionally need some manual downloads. 
And as the `needs-work` tag also implies we know that this is only a band-aid fix that will be worked on once we can figure out why the issue people with manual downloads had is happening and people have in general more time to work on the app and debug the issue this is mitigating.

Also if it gets to bad you can just restart WJ and it will analyze all the files you have already downloaded.
And with people being expected not download a list manually in a few hours it is expected that people will also close the app.

Or you get nexus premium where WJ only needs to open webview windows for some manual external sources and LoversLab.

If you think you can fix the issue that this is mitigating we are very open to welcome your contributions.

Also if you want to read more on why this change had to be made refer to this post and project update on the projects Patreon:
https://www.patreon.com/posts/93180808

**Comment by Nostras** [comment]
This is a problem for me because I use a bot that automatically clicks the download button for me whilst I have a stream open on my other screen where I watch a stream or something. Because the VRAM gets full my video/stream crashes and effectively freezes the system. It's quite surprising that this is even possible at all honestly, god bless Windows/AMD/whoever is supposed to properly deal with it.

Funnily enough I have had the same issues @EzioTheDeadPoet refers to in the Patreon project some time back, but I thought it was the nexusmods themselves doing a funny. You could argue that by doing this you're effectively making Wabbajack guaranteed to be the issue instead of users thinking it's, well, what I thought.

As a developer myself it pains me to see it being treated like this, I primarily have an issue with the 'not-a-bug' tag, because frankly both options I'd consider "bugged". If you're going to let this be a ~~bug~~ issue I'd strongly recommend coding in a reference to this topic, the patreon post and/or something that describes this issue whenever a user is going to download 50* mods or more from nexusmods and does not have premium.

Not to sound ungrateful, I'm really glad this tool exists, but _please_ inform the user about this issue instead of what happened here.

**Comment by Althro** [comment]
You do realize your clicker is against Nexus terms of service....

**Comment by Nostras** [comment]
> You do realize your clicker is against Nexus terms of service....

Oh yeah I am very well aware. I, however, do not promote it nor condone it. I heard of it yes. But that's not of your or Wabbajacks concern.

**Comment by EzioTheDeadPoet** [comment]
> As a developer myself it pains me to see it being treated like this, I primarily have an issue with the 'not-a-bug' tag, because frankly both options I'd consider "bugged". If you're going to let this be a ~bug~ issue I'd strongly recommend coding in a reference to this topic, the patreon post and/or something that describes this issue whenever a user is going to download 50* mods or more from nexusmods and does not have premium. A popup and an 'I agree' button should be dandy.
> 
> Not to sound ungrateful, I'm really glad this tool exists, but _please_ inform the user about this issue instead of what happened here.

You are free to add functionality for this in your free time. We do welcome external contributions.

**Comment by Nostras** [comment]
> > As a developer myself it pains me to see it being treated like this, I primarily have an issue with the 'not-a-bug' tag, because frankly both options I'd consider "bugged". If you're going to let this be a ~bug~ issue I'd strongly recommend coding in a reference to this topic, the patreon post and/or something that describes this issue whenever a user is going to download 50* mods or more from nexusmods and does not have premium. A popup and an 'I agree' button should be dandy.
> > Not to sound ungrateful, I'm really glad this tool exists, but _please_ inform the user about this issue instead of what happened here.
> 
> You are free to add functionality for this in your free time. We do welcome external contributions.

Sadly I'm not that kind of programmer, C# is alien to me and I especially do not develop against Windows myself.
Perhaps I can take a look some time, but a language one is not experienced in in a project this size is... unlikely to yield a good result in _reasonable_ effort for me.

**Comment by tr4wzified** [comment]
This is definitely a bug, it was a temporary workaround. Got a fixed version if you guys are interested: https://github.com/wabbajack-tools/wabbajack/pull/2478

**Comment by tr4wzified** [comment]
Link to a compiled release: https://mega.nz/file/sTRCAarR#k1rLmGrk5xA084L-fA2aid3pJWVm3YDL1KupbAxoMkE

**Comment by tr4wzified** [comment]
Fixed in 3.4.2.0+
